src/runtime/hexagon_remote/elf.h

```diff
@@ -0,0 +1,901 @@
+        *sec_text,     // The .text section where the functions live
+        *sec_strtab;   // The string table, for looking up symbol names
+
+    // The writeable portions of the object file in memory
+    char *writeable_buf;
+    size_t writeable_size;
+
+    // The global offset table for PIC code
+    elfaddr_t *global_offset_table;
+    int global_offset_table_entries;
+    int max_global_offset_table_entries;
+
+    // Load an object file in memory. Does not take
+    // ownership of the memory. Memory should be page-aligned.
+    void parse_object_file(const unsigned char *b, size_t s, bool d = false) {
+        failed = 0;
+        buf = NULL;
+        writeable_buf = NULL;
+        writeable_size = 0;
+        size = (s + alignment) & ~alignment;
+        debug = d;
+
+        int global_offset_table_size = 4096;
+
+        // Make a mapping of the appropriate size and type. We
+        // allocate the size of the object file for the executable
+        // stuff, the same size again to make a writeable copy, and
+        // then an extra page for the global offset table for PIC
+        // code.
+        buf = (char *)mmap(NULL, size * 2 + global_offset_table_size,
+                           PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
+
+        // Copy over the data
+        memcpy(buf, b, s);
+
+        // Set up the global offset table
+        global_offset_table = (elfaddr_t *)(buf + size*2);
+        global_offset_table_entries = 0;
+        max_global_offset_table_entries = global_offset_table_size / sizeof(elfaddr_t);
+        memset(global_offset_table, 0, global_offset_table_size);
+
+        // Grab the ELF header
+        if (size < sizeof(elf_header_t)) {
+            fail(__LINE__);
+            return;
+        }
+        header = (elf_header_t *)buf;
+
+        // Get the section names section first
+        sec_secnames = get_section(header->e_shstrndx);
+        if (failed) return;
+
+        // Walk over the other sections
+        for (int i = 0; i < num_sections(); i++) {
+            section_header_t *sec = get_section(i);
+            const char *sec_name = get_section_name(sec);
+            if (failed) return;
+            if (debug) log_printf("\nSection %s at %p:\n",
+                              sec_name,
+                              get_addr(get_section_offset(sec)));
+
+            if (sec->sh_type == 2) {
+                sec_symtab = sec;
+            } else if (sec->sh_type == 3) {
+                sec_strtab = sec;
+            } else if (sec->sh_type == 1 && strncmp(sec_name, ".text", 5) == 0) {
+                sec_text = sec;
+            }
+        }
+    }
+
+    void fail(int line) {
+        log_printf("Failure at line %d\n", line);
+        failed = line;
+    }
+
+    void move_writeable_sections() {
+        // Move the writeable sections to their own mapping. First
+        // determine their span.
```

Comment:
Is it possible for there to be an object where the first and last sections are writable, so we copy the whole object? It's fine if so for now, but it would be nice to put a comment in here that is either a TODO to improve this, or explain why that isn't the case.